### PR TITLE
Check for non-active/supported cgroups

### LINKED ISSUE #76: unable to find * in controller set: unknown
There are problems with this library handling missing cgroups or the way they are mounted. The unhandled configuration seems to become common with new systemd.

These issues list the output of mount and proc in the affected systems:

https://github.com/docker/for-linux/issues/545
https://github.com/docker/for-linux/issues/552

@crosbymichael

### PR DESCRIPTION
**Pull request by crosbymichael**
Fixes #76

I still need to get a system setup to test this fix.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

### CODE REVIEW

**Comment by crosbymichael** [comment]
I was able to verify this on a system that Tonis setup for me.  I'm happy with these changes and this is good to review.

```bash
>  docker run alpine true
docker: Error response from daemon: controller is not supported: unknown.
ERRO[0000] error waiting for container: context canceled

>  cp * /usr/bin/ #patched containerd
> docker run alpine true
> docker run alpine ls
bin
dev
etc
home
lib
media
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
```

File: cgroup.go [hunk 1]
```diff
@@ -60,6 +77,16 @@ func Load(hierarchy Hierarchy, path Path) (Cgroup, error) {
 			if os.IsNotExist(errors.Cause(err)) {
 				return nil, ErrCgroupDeleted
 			}
+			if err == ErrControllerNotActive {
```
Comment (Ace-Tang) [inline cgroup.go]: Do we need to change the location for `if os.IsNotExist(errors.Cause(err))` and  `if err == ErrControllerNotActive`, like
```
			if err == ErrControllerNotActive {
				for _, o := range skip {
					if skerr := o(s, path, err); skerr != nil {
						if skerr != ErrSkipSubsystem {
							return nil, skerr
						}
					}
				}
				continue
			}
                        if os.IsNotExist(errors.Cause(err)) {
				return nil, ErrCgroupDeleted
			}
```
first check controller is support, then check the file is exist ?

**Comment by Ace-Tang** [comment]
Overall looks really good to me, from reading the patch, I think it can solve my problem (I mount some controller like pid, cpu, memory through cgroup2)

File: cgroup.go (hunk 1 above)
Comment (crosbymichael) [reply cgroup.go]: I think this is fine to keep the existing logics for Load, the isnotexist is for when the cgroup directory structure has been removed, not an active controller.  I think this is fine for the Load case to get the ErrCgroupDeleted error which is handled

File: cgroup.go [hunk 2]
```diff
@@ -30,24 +30,41 @@ import (
...
+				continue
+			}
 			return nil, err
 		}
+		active = append(active, s)
 	}
 	return &cgroup{
 		path:       path,
-		subsystems: subsystems,
+		subsystems: active,
```
Comment (mikebrow) [inline cgroup.go]: how about adding a list for skipped .. or inactive?

File: cgroup.go [hunk 3]
```diff
@@ -30,24 +30,41 @@ import (
...
 	}, nil
 }
 
 // Load will load an existing cgroup and allow it to be controlled
-func Load(hierarchy Hierarchy, path Path) (Cgroup, error) {
+func Load(hierarchy Hierarchy, path Path, skip ...SkipOpts) (Cgroup, error) {
```
Comment (mikebrow) [inline cgroup.go]: ditto for update to function description

File: paths.go [hunk 4]
```diff
@@ -57,6 +57,9 @@ func PidPath(pid int) Path {
 	return existingPath(paths, "")
 }
 
+// ErrControllerNotActive is returned when a controller is not supported or enalbed
```
Comment (mlaventure) [inline paths.go]: `s/enalbed/enabled/`

File: cgroup.go (hunk 2 above)
Comment (crosbymichael) [reply cgroup.go]: not sure we need it

**Pull request merged**